File: app/src/main/java/com/iogarage/ke/pennywise/TransactionAdapter.java
package com.iogarage.ke.pennywise;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Color;
import androidx.core.content.ContextCompat;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageView;
import android.widget.TextView;

import com.iogarage.ke.pennywise.api.ContactManager;
import com.iogarage.ke.pennywise.entities.Debt;
import com.iogarage.ke.pennywise.util.DateUtil;
import com.iogarage.ke.pennywise.util.Util;

import org.greenrobot.eventbus.EventBus;
import org.threeten.bp.LocalDate;
import org.threeten.bp.temporal.ChronoUnit;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.LinkedList;
import java.util.List;

/**
 * Created by Joshua on 11/1/2014.
 */
public class TransactionAdapter extends RecyclerView.Adapter<TransactionAdapter.ViewHolder> implements Filterable {
    private List<Debt> _dataset;
    private List<Debt> _original;



    private Context mContext;

    public TransactionAdapter(List<Debt> dataset, Context context) {
        _dataset = dataset;
        _original = new LinkedList<>(dataset);
        mContext = context;
    }

    @Override
    public Filter getFilter() {
        return new TransactionFilter(this, _original);
    }

    public void setData(List<Debt> dataset) {
        _dataset = dataset;
        _original = new LinkedList<>(dataset);
        notifyDataSetChanged();
    }

    public void remove(int position) {
        if (!_dataset.isEmpty()) {


            Debt t = _dataset.get(position);
            EventBus.getDefault().post(new DeleteTransaction(t));

            _dataset.remove(position);
            _original.remove(position);
            notifyDataSetChanged();


        }

    }

    public void selected(int position) {
        Debt transaction = _dataset.get(position);
        Intent intent = new Intent(mContext, PaymentView.class);
        intent.putExtra(PaymentView.TRANSACTION_ID, transaction.getId());
        mContext.startActivity(intent);
    }

    // Not use static
    public class ViewHolder extends RecyclerView.ViewHolder  implements View.OnClickListener{
        public TextView mTextView;
        public TextView mTrxDate;
        public TextView mAmount;
        public ImageView image;
        public TextView mCurrency;
        public TextView status;
        public TextView textSeparator;
        CardView transactionCard;

        public ViewHolder(View itemView) {
            super(itemView);
            mTextView = itemView.findViewById(R.id.txtname);
            mTrxDate = itemView.findViewById(R.id.txtenddate);
            mAmount = itemView.findViewById(R.id.txtamount);
            mCurrency = itemView.findViewById(R.id.txtcurrency);
            image = itemView.findViewById(R.id.imageView);
            status = itemView.findViewById(R.id.status);
            textSeparator  = itemView.findViewById(R.id.header_separator);
            transactionCard = itemView.findViewById(R.id.transaction_card);

            transactionCard.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {

            if(v.getId() == R.id.header_separator)
                return;

            int position  =   getAdapterPosition();

            selected(position);
        }
    }

    @Override
    public int getItemCount() {
        return _dataset.size();
    }

    @Override
    public void onBindViewHolder(ViewHolder holder,  int position) {

        final Debt transaction = _dataset.get(position);

        Calendar calendar = Calendar.getInstance();
        calendar.setTime(transaction.getPaydate());

        // Show header for item if it is the first in date group
        if (position > 0 && transaction.getPaydate().equals(_dataset.get(position - 1).getPaydate()) ) {
            holder.textSeparator.setVisibility(View.GONE);
        } else {
            String appropriateDate = DateUtil.getAppropriateDateFormat(mContext, calendar);
            holder.textSeparator.setText(appropriateDate);
            holder.textSeparator.setVisibility(View.VISIBLE);
        }

        holder.mTextView.setText(transaction.getPersonname());

        if (transaction.getType() == TransactionView.LENDING) {
            holder.mAmount.setTextColor(ContextCompat.getColor(mContext, R.color.md_red_500));
            holder.mAmount.setText(Util.formatCurrency(transaction.getBalance()));
        } else {
            holder.mAmount.setTextColor(ContextCompat.getColor(mContext, R.color.md_green_500));
            holder.mAmount.setText(Util.formatCurrency(transaction.getBalance()));
        }
        holder.mCurrency.setText(transaction.getCurrency());
        holder.mTrxDate.setText("Due date: " + Util.formatDate(transaction.getPaydate()));

        Bitmap photo = null;

        try {
            photo = ContactManager.getPhoto(mContext, transaction.getPhonenumber());
        } catch (Exception ex) {
            ex.printStackTrace();
        }

        if (photo != null) {
            //Bitmap photo = BitmapFactory.decodeByteArray(transaction.getImage(), 0, transaction.getImage().length);
            holder.image.setImageBitmap(photo);
        }

        LocalDate today  = LocalDate.now();
        LocalDate toDate = DateUtil.date2LocalDate(transaction.getPaydate());

        long days = ChronoUnit.DAYS.between(today, toDate);

        if (days > 1)
            holder.status.setText(days + " days");
        else if (days == 1)
            holder.status.setText(days + " day");
        else if (days == 0) {
            holder.status.setText("due");
            holder.status.setTextColor(Color.RED);
        } else {
            holder.status.setText("overdue");
            holder.status.setTextColor(Color.RED);
        }

        if (transaction.getBalance() <= 0)
            holder.status.setText("paid");





    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(
                R.layout.item_transactions_list, parent, false);
        ViewHolder holder = new ViewHolder(view);
        return holder;
    }

    private static class TransactionFilter extends Filter {

        private final TransactionAdapter adapter;

        private final List<Debt> originalList;

        private final List<Debt> filteredList;

        private TransactionFilter(TransactionAdapter adapter, List<Debt> originalList) {
            super();
            this.adapter = adapter;
            this.originalList = new LinkedList<>(originalList);
            this.filteredList = new ArrayList<>();
        }

        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            filteredList.clear();
            final FilterResults results = new FilterResults();

            if (constraint.length() == 0) {
                filteredList.addAll(originalList);
            } else {
                final String filterPattern = constraint.toString().toLowerCase().trim();

                for (final Debt transaction : originalList) {
                    String name = transaction.getPersonname().toLowerCase().trim();
                    String phone = transaction.getPhonenumber().toLowerCase().trim();

                    if (name.startsWith(filterPattern) || phone.startsWith(filterPattern)) {
                        filteredList.add(transaction);
                    }
                }
            }
            results.values = filteredList;
            results.count = filteredList.size();
            return results;
        }

        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            adapter._dataset.clear();
            adapter._dataset.addAll((ArrayList<Debt>) results.values);
            adapter.notifyDataSetChanged();
        }
    }
}

File: app/src/main/java/com/iogarage/ke/pennywise/views/transactions/TransactionAdapter.kt
package com.iogarage.ke.pennywise.views.transactions

import android.graphics.Bitmap
import android.graphics.Color
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import android.widget.ImageView
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.iogarage.ke.pennywise.R
import com.iogarage.ke.pennywise.domain.entity.Transaction
import com.iogarage.ke.pennywise.domain.entity.TransactionType
import com.iogarage.ke.pennywise.util.Util
import com.iogarage.ke.pennywise.util.asString
import java.time.LocalDate
import java.time.temporal.ChronoUnit
import java.util.LinkedList
import java.util.Locale

/**
 * Created by Joshua on 11/1/2014.
 */
class TransactionAdapter(
    private var _dataset: List<Transaction>,
    private val onItemSelected: (Long) -> Unit
) : RecyclerView.Adapter<TransactionAdapter.ViewHolder>(), Filterable {
    private var transactions: MutableList<Transaction> = _dataset.toMutableList()

    override fun getFilter(): Filter {
        return TransactionFilter(this, transactions)
    }

    fun setData(dataset: List<Transaction>) {
        transactions = dataset.toMutableList()
        notifyDataSetChanged()
    }

    fun remove(position: Int) {
        if (_dataset.isNotEmpty()) {
            transactions.removeAt(position)
            notifyDataSetChanged()
        }
    }

    fun selected(position: Int) {
        val item = _dataset[position]
        onItemSelected(item.transactionId)
    }

    // Not use static
    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView),
        View.OnClickListener {
        var mTextView: TextView
        var mTrxDate: TextView
        var mAmount: TextView
        var image: ImageView
        var mCurrency: TextView
        var status: TextView
        var textSeparator: TextView
        private var transactionCard: CardView

        init {
            mTextView = itemView.findViewById(R.id.txtname)
            mTrxDate = itemView.findViewById(R.id.txtenddate)
            mAmount = itemView.findViewById(R.id.txtamount)
            mCurrency = itemView.findViewById(R.id.txtcurrency)
            image = itemView.findViewById(R.id.imageView)
            status = itemView.findViewById(R.id.status)
            textSeparator = itemView.findViewById(R.id.header_separator)
            transactionCard = itemView.findViewById(R.id.transaction_card)
            transactionCard.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            if (v.id == R.id.header_separator) return
            val position = adapterPosition
            selected(position)
        }
    }

    override fun getItemCount(): Int {
        return _dataset.size
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val (_, _, personName, _, _, _, payDate, _, balance, type, currency) = _dataset[position]
        val date = LocalDate.ofEpochDay(payDate)
        val debug = date.asString("MMM dd", Locale.getDefault())
        Log.d("END_DATE", "$debug $payDate")

        // Show header for item if it is the first in date group
        if (position > 0 && payDate == _dataset[position - 1].payDate) {
            holder.textSeparator.visibility = View.GONE
        } else {
            val header = date.asString(
                "MMM dd",
                Locale.getDefault()
            ) // getAppropriateDateFormat(mContext, calendar);
            holder.textSeparator.text = header
            holder.textSeparator.visibility = View.VISIBLE
        }
        holder.mTextView.text = personName
        if (type === TransactionType.LENDING) {
            holder.mAmount.setTextColor(
                ContextCompat.getColor(
                    holder.itemView.context,
                    R.color.md_red_500
                )
            )
            holder.mAmount.text = Util.formatCurrency(balance)
        } else {
            holder.mAmount.setTextColor(
                ContextCompat.getColor(
                    holder.itemView.context,
                    R.color.md_green_500
                )
            )
            holder.mAmount.text = Util.formatCurrency(balance)
        }
        holder.mCurrency.text = currency
        val dueDate = date.asString(
            "E, dd MMM, yyyy",
            Locale.getDefault()
        ) // getAppropriateDateFormat(mContext, calendar);
        holder.mTrxDate.text = "Due date: $dueDate"
        val photo: Bitmap? = null
        try {
            //  photo = ContactManager.getPhoto(mContext, transaction.getPhonenumber());
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
        if (photo != null) {
            //Bitmap photo = BitmapFactory.decodeByteArray(transaction.getImage(), 0, transaction.getImage().length);
            holder.image.setImageBitmap(photo)
        }
        val today = LocalDate.now()
        val toDate = LocalDate.ofEpochDay(payDate)
        val days = ChronoUnit.DAYS.between(today, toDate)
        if (days > 1) holder.status.text = "$days days" else if (days == 1L) holder.status.text =
            "$days day" else if (days == 0L) {
            holder.status.text = "due"
            holder.status.setTextColor(Color.RED)
        } else {
            holder.status.text = "overdue"
            holder.status.setTextColor(Color.RED)
        }
        if (balance <= 0) holder.status.text = "paid"
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(
            R.layout.item_transactions_list, parent, false
        )
        return ViewHolder(view)
    }

    private class TransactionFilter constructor(
        private val adapter: TransactionAdapter,
        originalList: List<Transaction>
    ) : Filter() {
        private val _originalList: List<Transaction>
        private val filteredList: MutableList<Transaction>

        init {
            _originalList = LinkedList(originalList)
            filteredList = ArrayList()
        }

        override fun performFiltering(constraint: CharSequence): FilterResults {
            filteredList.clear()
            val results = FilterResults()
            if (constraint.isEmpty()) {
                filteredList.addAll(_originalList)
            } else {
                val filterPattern =
                    constraint.toString().lowercase(Locale.getDefault()).trim { it <= ' ' }
                for ((transactionId, transactionDate, personName, phoneNumber, amount, note, payDate, paid, balance, type, currency, status, reminderTitle, reminderContent, reminderDate, reminderStatus) in _originalList) {
                    /*  String name = transaction.getPersonname().toLowerCase().trim();
                    String phone = transaction.getPhonenumber().toLowerCase().trim();

                    if (name.startsWith(filterPattern) || phone.startsWith(filterPattern)) {
                        filteredList.add(transaction);
                    }*/
                }
            }
            results.values = filteredList
            results.count = filteredList.size
            return results
        }

        override fun publishResults(constraint: CharSequence, results: FilterResults) {
            adapter.transactions.clear()
            adapter.transactions.addAll((results.values as ArrayList<Transaction>))
            adapter.notifyDataSetChanged()
        }
    }
}